packages/plugin-ext/src/main/browser/plugin-authentication-service.ts

```diff
@@ -0,0 +1,70 @@
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { AuthenticationProvider, AuthenticationService, AuthenticationServiceImpl, AuthenticationSession } from '@theia/core/lib/browser/authentication-service';
+import { inject } from '@theia/core/shared/inversify';
+import { HostedPluginSupport } from '../../hosted/browser/hosted-plugin';
+
+export function getAuthenticationProviderActivationEvent(id: string): string { return `onAuthenticationRequest:${id}`; }
+
+/**
+ * Plugin authentiction service that aims to activate additional plugins if sessions are created or queried.
+ */
+export class PluginAuthenticationServiceImpl extends AuthenticationServiceImpl implements AuthenticationService {
+
+    constructor(@inject(HostedPluginSupport) protected readonly pluginService: HostedPluginSupport) {
+        super();
+    }
+
+    async getSessions(id: string, scopes?: string[]): Promise<ReadonlyArray<AuthenticationSession>> {
+        await this.tryActivateProvider(id);
+        return super.getSessions(id, scopes);
+    }
+
+    async login(id: string, scopes: string[]): Promise<AuthenticationSession> {
+        await this.tryActivateProvider(id);
+        return super.login(id, scopes);
+    }
+
+    private async tryActivateProvider(providerId: string): Promise<AuthenticationProvider> {
+        this.pluginService.activateByEvent(getAuthenticationProviderActivationEvent(providerId));
+
+        let provider = this.authenticationProviders.get(providerId);
+        if (provider) {
+            return provider;
+        }
+
+        // When activate has completed, the extension has made the call to `registerAuthenticationProvider`.
+        // However, activate cannot block on this, so the renderer may not have gotten the event yet.
+        const didRegister: Promise<AuthenticationProvider> = new Promise((resolve, _) => {
+            this.onDidRegisterAuthenticationProvider(info => {
+                if (info.id === providerId) {
+                    provider = this.authenticationProviders.get(providerId);
+                    if (provider) {
+                        resolve(provider);
+                    } else {
+                        throw new Error(`No authentication provider '${providerId}' is currently registered.`);
+                    }
+                }
+            });
+        });
+
+        const didTimeout: Promise<AuthenticationProvider> =
+            new Promise((_, reject) => setTimeout(() => reject('Timed out waiting for authentication provider to register'), 5000));
```

Comment:
Suggestion: We already have a `timeout` function in the `promise-util` which works quite similar. But instead of rejecting the promise, the `timeout` function resolves it. What do you think about adding a `timeoutReject` method which looks something like this?

```ts
export function timeoutReject(ms: number, message?: string): Promise<void> {
    const deferred = new Deferred<void>();
    setTimeout(() => deferred.reject(new Error(message)), ms);
    return deferred.promise;
}
```

This could also be used [here](https://github.com/eclipse-theia/theia/blob/37ed897594a709d7e0a10d3db4c3bba85f00878a/packages/monaco/src/browser/monaco-editor-provider.ts#L307).

I also thought about an additional override for `timeout`, but switching whether the promise resolves or rejects depending on an optional argument seems to me like a weird change of semantics.

Your opinion on that?

Reply:
I think that is a great idea! I was actually looking for a function like that but then simply decided to do it like in the vscode code. I took the method that you proposed since I am also not a huge fan of a Boolean parameter to switch between resolution and rejection since I know that I always have to double and triple check Boolean parameters when reading/writing to ensure correct usage ;-)